proposed/2021/MachineReadableOutput.md

```diff
@@ -0,0 +1,601 @@
+```json
+{
+  "version": 1,
+   "sources": [
+    "https://api.nuget.org/v3/index.json",
+    "https://apidev.nugettest.org/v3-index/index.json"
+  ], 
+  "MyProjectA": [
+    {
+      "framework": "netcoreapp3.1",
+      "topLevelPackages": [
+        {
+          "id": "Microsoft.Extensions.Primitives",
+          "requestedVersion": "[1.0.0, 5.0.0]",
+          "resolvedVersion": "1.0.0",
+          "latestVersion": "6.0.0",
+        },
+        {
+          "id": "NuGet.Commands",
+          "requestedVersion": "4.8.0-preview3.5278",
+          "resolvedVersion": "4.8.0-preview3.5278",
+          "latestVersion": "6.0.0",
+        },
+        {
+          "id": "Text2Xml.Lib",
+          "requestedVersion": "[1.1.2, 2.0.0)",
+          "resolvedVersion": "1.1.2",
+          "latestVersion": "1.1.4"
+        }
+      ]
+    }
+  ],
+  "MyProjectB": [
+    {
+      "framework": "netcoreapp3.1",
+      "topLevelPackages": [
+        {
+          "id": "NuGet.Commands",
+          "requestedVersion": "4.8.0-preview3.5278",
+          "resolvedVersion": "4.8.0-preview3.5278",
+          "latestVersion": "6.0.0"
+        },
+        {
+          "id": "Text2Xml.Lib",
+          "requestedVersion": "1.1.2",
+          "resolvedVersion": "1.1.2",
+          "latestVersion": "1.1.4"
+        }
+      ]
+    },
+    {
+      "framework": "net5.0",
+      "topLevelPackages": [ 
+        {
+          "id": "NuGet.Commands",
+          "requestedVersion": "4.8.0-preview3.5278",
+          "resolvedVersion": "4.8.0-preview3.5278",
+          "latestVersion": "6.0.0"
+        },
+        {
+          "id": "Text2Xml.Lib",
+          "requestedVersion": "1.1.2",
+          "resolvedVersion": "1.1.2",
+          "latestVersion": "1.1.4"
+        }
+      ]
+    }
+  ]
+}
+```
+
+#### `> dotnet list package --deprecated`
+
+```dotnetcli
+The following sources were used:
+   https://api.nuget.org/v3/index.json
+   https://apidev.nugettest.org/v3-index/index.json
+
+Project `MyProjectA` has the following deprecated packages
+   [netcoreapp3.1]:
+   Top-level Package                 Requested   Resolved   Reason(s)   Alternative
+   > EntityFramework.MappingAPI      *           6.2.1      Legacy      Z.EntityFramework.Extensions >= 0.0.0
+   > NuGet.Core                      2.13.0      2.13.0     Legacy
+
+Project `MyProjectB` has the following deprecated packages
+   [netcoreapp3.1]:
+   Top-level Package      Requested   Resolved   Reason(s)   Alternative
+   > NuGet.Core           2.13.0      2.13.0     Legacy
+
+   [net5.0]:
+   Top-level Package      Requested   Resolved   Reason(s)   Alternative
+   > NuGet.Core           2.13.0      2.13.0     Legacy
+```
+
+#### `> dotnet list package --deprecated --format json`
+
+```json
+{
+  "version": 1,
+   "sources": [
+    "https://api.nuget.org/v3/index.json",
+    "https://apidev.nugettest.org/v3-index/index.json"
+  ],
+  "MyProjectA": [
+    {
+      "framework": "netcoreapp3.1",
+      "topLevelPackages": [
```

Comment:
How would this format evolve if we add another "package pivot" in addition to top level and transitive packages? For example, what if we add new package kinds for source generators, Roslyn analyzers, etc...?

What are your thoughts of merging `topLevelPackages` and `transitivePackages` into a single property, like `dependencies`? For example, before:

```json
"topLevelPackages": [
  {
    "id": "MyTop",
     ...
  }
],
"transitivePackages": [
  {
    "id": "MyTransitive",
     ...
  }
]
```

 After:

```json
"dependencies": [
  {
    "id": "MyTop",
    "transitive": false,
     ...
  },
  {
    "id": "MyTransitive",
    "transitive": true,
     ...
  }
],
```

Reply:
I have no strong opinion at this time. 
I believe `versioning` can help us with dealing changes. Please check thread on it  
[#r766750418](https://github.com/NuGet/Home/pull/11446/files#r766750418)

Reply:
What are your thoughts of merging `topLevelPackages` and `transitivePackages` into a single property?

Reply:
@loic-sharma 
At this moment I'm believe `json` output be "1:1" representation of `text` output, otherwise I would be hard to reason for broader usage case, of course I could be wrong. Currently I don't know how `source generators, Roslyn analyzers` play out. Maybe if it really evolves over time, then `versioning` should come into play.

Reply:
> At this moment I believe json output be "1:1" representation of text output

Why are we making this assumption? I feel like it will deliver features that customers don't want. Of what practical use is `dotnet list package --format json` without at least one of `--include-transitive`, `--deprecated`, `--vulnerable`, or `--outdated`?

As I brought up elsewhere in this PR, but I strongly believe that customers are going to want both deprecated and vulnerability information, if not outdated as well, in a single output without being forced to call the command 3 times. For anything that can be added later without introducing a break change, I fully accept we can deliver an MVP. But for anything that will require a breaking change, customers are going to use this like an API, not an interactive tool, so I disagree that it's acceptable to customers for us to break the output in future versions of the tool.

Reply:
> > At this moment I believe json output be "1:1" representation of text output
> 
> Why are we making this assumption? I feel like it will deliver features that customers don't want. Of what practical use is `dotnet list package --format json` without at least one of `--include-transitive`, `--deprecated`, `--vulnerable`, or `--outdated`?

> 
> As I brought up elsewhere in this PR, but I strongly believe that customers are going to want both deprecated and vulnerability information, if not outdated as well, in a single output without being forced to call the command 3 times. For anything that can be added later without introducing a break change, I fully accept we can deliver an MVP. But for anything that will require a breaking change, customers are going to use this like an API, not an interactive tool, so I disagree that it's acceptable to customers for us to break the output in future versions of the tool.

Merging them is it's possible design, currently I'm still not convinced why this design it better than current one. 
Nothing prevents customer doing `dotnet list package --include-transitive --vulnerable --format json`. 
If you are talking about include options by default then it's out of scope, I have tracking issue for it : https://github.com/NuGet/Home/issues/11550
, https://github.com/NuGet/Home/issues/11551
Also, this overlaps  with `dotnet audit` spec too.

Reply:
Discussed offline. Out of scope for now.